README.md
# Matomo Plugin

The **Matomo** Plugin is an extension for [Grav CMS](http://github.com/getgrav/grav). It integrates Matomo analytics into Grav CMS. Unlike other plugins this supports **server side tracking** via [Matomo PHP Tracking API](https://github.com/matomo-org/matomo-php-tracker). This plugin **requires no client side Javascript** but is easier to use compared to [Matomo Log Analysis](https://matomo.org/docs/log-analytics-tool-how-to/).

<a href="https://www.buymeacoffee.com/nicohood" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>

## Features

* Easy server-side tracking
* Matomo Dashboard integrated in Grav Admin Plugin
* Optional Cookie tracking
* Optional Javascript tracking
* Grav 1.6 and 1.7 supported

## Installation

Installing the Matomo plugin can be done in one of three ways: The GPM (Grav Package Manager) installation method lets you quickly install the plugin with a simple terminal command, the manual method lets you do so via a zip file, and the admin method lets you do so via the Admin Plugin.

### GPM Installation (Preferred)

To install the plugin via the [GPM](http://learn.getgrav.org/advanced/grav-gpm), through your system's terminal (also called the command line), navigate to the root of your Grav-installation, and enter:

    bin/gpm install matomo

This will install the Matomo plugin into your `/user/plugins`-directory within Grav. Its files can be found under `/your/site/grav/user/plugins/matomo`.

### Manual Installation

To install the plugin manually, download the zip-version of this repository and unzip it under `/your/site/grav/user/plugins`. Then rename the folder to `matomo`. You can find these files on [GitHub](https://github.com/nico-hood/grav-plugin-matomo) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/matomo

> NOTE: This plugin is a modular component for Grav which may require other plugins to operate, please see its [blueprints.yaml-file on GitHub](https://github.com/nico-hood/grav-plugin-matomo/blob/master/blueprints.yaml).

### Admin Plugin

If you use the Admin Plugin, you can install the plugin directly by browsing the `Plugins`-menu and clicking on the `Add` button.

## Requirements

This plugin uses the [Matomo PHP Tracking API](https://github.com/matomo-org/matomo-php-tracker) which depends on the following PHP modules:

- json extension (json_decode, json_encode)
- CURL or STREAM extensions (to issue the HTTPS request to Matomo)

## Configuration

Before configuring this plugin, you should copy the `user/plugins/matomo/matomo.yaml` to `user/config/plugins/matomo.yaml` and only edit that copy.

Here is the default configuration and an explanation of available options:

```yaml
enabled: true
active: true

# Connection settings
matomo_url: 'https://example.tld'
site_id: 1
# It is recommended to set the token via the grav dotenv plugin:
# https://github.com/Ralla/grav-plugin-dotenv
token: ''

# Privacy settings
respect_do_not_track: true
enable_cookies: false
enable_javascript: false
# Array of blocked client IP addresses for which tracking will be disabled.
blockedIpAddresses: []
# Array of blocked client IPv4 and/or IPv6 address ranges in the form
# ["192.177.204.1-192.177.204.254", "2001:db8::1-2001:db8::fe", ...].
# In addition to numerical ranges, the keywords "private", "loopback", "link-local" are recognized,
# designating special IPv4 and IPv6 ranges (see RFCs 6890, 4193, 4291).
blockedIpRanges: ["private", "loopback", "link-local"]
# Name of a blocking cookie, which disables Matomo JS and PHP tracking, when set.
blockingCookie: "blockMatomo"

# Grav admin plugin dashboard settings
# The dashboard token must only have read access:
# https://matomo.org/docs/embed-matomo-reports/#embed-piwik-widgets-on-a-password-protected-or-private-page
dashboard_token: ''
# Specify a different site id for the dashboard.
# This is useful if you have a test and production environment
# and always want to show the production stats in the admin panel.
dashboard_site_id: ''

# Development settings
# Enable this to add debug output about blocking reasons to the page's HTML source ('Matomo tracking blocked').
debug: false
```

Note that if you use the Admin Plugin, a file with your configuration named matomo.yaml will be saved in the `user/config/plugins/`-folder once the configuration is saved in the Admin.

It is **recommended to set the token via environment variables** and the [grav dotenv plugin](https://github.com/Ralla/grav-plugin-dotenv). Please **be aware** that the token will be currently [added to the config file](https://github.com/Ralla/grav-plugin-dotenv/issues/11) when editing the settings via the admin plugin.

## Usage

Simply configure your matomo host in the config and you are ready to go! Please note, that by default this plugin is **configured privacy focussed**. You and enable additional features like tracking with cookies, javascript or ignoring the "Do Not Track" header in the config. **Note:** Ignoring the "Do Not Track" header only makes sense when using the php tracker, not the javascript code. That's because the matomo instance will also evaluate the header and ignore the request.

The default config will only track the visited url, ip, referrer, language, browser/user agent, country (via ip lookup, if enabled on matomo) which is similar to [Matomo Log Analysis](https://matomo.org/docs/log-analytics-tool-how-to/), just that this plugin is more easy to use, extensible and gives instant results.

If you have any improvement, please do not hesitate to file an issue.

## Credits

* [Matomo PHP Tracking API](https://github.com/matomo-org/matomo-php-tracker)

## To Do

- [ ] Track events like form submissions with an additional form action
- [ ] Add twig templates/css to also track impressions of objects via tracking pixel and mouseover



CHANGELOG.md
# v1.3.0

##  03/06/2021

1. [](#new)
   
    * Log php tracker errors instead of failing hard.
    
    * Add support for client IP address blocking and a blocking cookie #2

# v1.2.1

##  12/31/2020

1. [](#bugfix)
    * Fixed language string to support Grav 1.7

# v1.2.0
##  12/23/2020

1. [](#new)
    * Added Grav Admin Plugin Dashboard integration

# v1.1.0
##  12/23/2020

1. [](#new)
    * Added javascript tracking support

# v1.0.0
##  12/22/2020

1. [](#new)
    * Initial release


matomo.php
<?php
namespace Grav\Plugin;

use Composer\Autoload\ClassLoader;
use Grav\Common\Plugin;
use RocketTheme\Toolbox\Event\Event;
use MatomoTracker;

/**
 * Class MatomoPlugin
 * @package Grav\Plugin
 */
class MatomoPlugin extends Plugin
{
    /**
     * @return array
     *
     * The getSubscribedEvents() gives the core a list of events
     *     that the plugin wants to listen to. The key of each
     *     array section is the event that the plugin listens to
     *     and the value (in the form of an array) contains the
     *     callable (or function) as well as the priority. The
     *     higher the number the higher the priority.
     */
    public static function getSubscribedEvents(): array
    {
        return [
            'onPluginsInitialized' => [
                ['autoload', 100000], // TODO: Remove when plugin requires Grav >=1.7
                ['onPluginsInitialized', 0]
            ]
        ];
    }

    /**
    * Composer autoload.
    *is
    * @return ClassLoader
    */
    public function autoload(): ClassLoader
    {
        return require __DIR__ . '/vendor/autoload.php';
    }

    /**
     * Initialize the plugin
     */
    public function onPluginsInitialized(): void
    {
        // Don't proceed if we are in the admin plugin
        if ($this->isAdmin()) {
            if ($this->grav['config']->get('plugins.matomo.dashboard_token')) {
                $this->enable([
                    'onTwigTemplatePaths' => ['onTwigAdminTemplatePaths', 0],
                    'onAdminMenu' => ['onAdminMenu', 0]
                ]);
            }

            return;
        }

        // Enable the main events we are interested in
        $this->enable([
            'onPageInitialized' => ['onPageInitialized', 0],
            'onTwigTemplatePaths' => ['onTwigTemplatePaths', 0]
        ]);
    }

    /**
     * Add plugin templates path
     */
    public function onTwigAdminTemplatePaths()
    {
        $this->grav['twig']->twig_paths[] = __DIR__ . '/admin/templates';
    }

    /**
     * Add navigation item to the admin plugin
     */
    public function onAdminMenu()
    {
        $this->grav['twig']->plugins_hooked_nav['Matomo'] = ['route' => 'matomo', 'icon' => 'fa-bar-chart'];
    }

    public function onPageInitialized(Event $event)
    {
        $page = $event['page'];

        // Merge configs and then check for the active flag per page
        $config = $this->mergeConfig($page);
        if (!$config->get('active', true)) {
            $this->documentBlockingReason("Plugin is not active for this page.");
            return;
        }

        // Check if client set "Do Not Track" header
        // NOTE: Ignoring the header only works for php tracking, not for javascript tracking.
        // That's because the matomo instance will also evaluate the header and ignore the request.
        if ($config->get('respect_do_not_track', true)) {
            if (isset($_SERVER['HTTP_DNT']) && $_SERVER['HTTP_DNT'] == 1) {
                $this->documentBlockingReason("Client requests 'DO NOT TRACK'.");
                return;
            }
        }

        // Don't proceed if a blocking cookie is set
        $blockingCookieName = $config->get('blockingCookie');
        if (!empty($blockingCookieName) && !empty($_COOKIE[$blockingCookieName])) {
            $this->documentBlockingReason("Blocking cookie \"$blockingCookieName\" is set.");
            return;
        }

        // Don't proceed if the IP address is blocked
        if (in_array($_SERVER['REMOTE_ADDR'], $config->get('blockedIpAddresses', []))) {
            $this->documentBlockingReason("Client ip " . $_SERVER['REMOTE_ADDR'] . " is in blockedIps.");
            return;
        }

        // Don't proceed if the IP address is within a blocked range
        foreach ($config->get('blockedIpRanges', []) as $blockedIpRange) {
            if ($this->inIPAddressRange($this->packedIPAddress($_SERVER['REMOTE_ADDR']), $blockedIpRange)) {
                $this->documentBlockingReason("Client ip " . $_SERVER['REMOTE_ADDR'] . " is in range \"" . $blockedIpRange . "\".");
                return;
            }
        }

        $matomo_url = $config->get('matomo_url');
        $site_id = $config->get('site_id');
        $token = $config->get('token');

        if (!$matomo_url || !$site_id || !$token || $matomo_url === 'https://example.tld') {
            $this->documentBlockingReason('Invalid Matomo configuration detected.');
            return;
        }

        // Add javascript tracking code (and disable php tracker to avoid duplicate entries)
        if ($config->get('enable_javascript', false)) {
            $matomo_js = $this->grav['twig']->processTemplate('js/matomo.js.twig');
            $this->grav['assets']->addInlineJs($matomo_js,
            [
              'type' => 'text/javascript',
              'priority' => 0,
              'position' => 'after'
            ]);
            return;
        }

        // Matomo object
        $matomoTracker = new MatomoTracker((int)$site_id, $matomo_url);

        // Optionally enable cookies
        if (!$config->get('enable_cookies', false)) {
            $matomoTracker->disableCookieSupport();
        }

        // Set authentication token
        $matomoTracker->setTokenAuth($token);

        // Track page view
        // NOTE: Url, Ip, Referrer, Browser Language and UserAgent is set by the API automatically
        // TODO return value is currently useless: https://github.com/matomo-org/matomo-php-tracker/issues/85
        try {
            $ret = $matomoTracker->doTrackPageView($page->title());
        }
        catch (\RuntimeException $e) {
            $this->grav['log']->error('Error tracking page view via Matomo: ' . $e->getMessage());
        }
    }

    /**
     * Add templates directory to twig lookup paths.
     */
    public function onTwigTemplatePaths()
    {
        $this->grav['twig']->twig_paths[] = __DIR__ . '/templates';
    }

    /**
     * Documents the reason for blocking Matomo tracking in a JavaScript comment
     * @param string $reason
     */
    private function documentBlockingReason(string $reason)
    {
        if ($this->config->get('plugins.matomo.debug', false)) {
            $this->grav['assets']->addInlineJs("/* Matomo tracking blocked, reason: $reason */");
        }
    }

    /**
     * Returns a packed IP address which can be directly compared to another packed IP address
     * @param string $humanReadableIPAddress IPv4 or IPv6 address in human-readable notation
     * @return string (16 byte packed representation)
     */
    private function packedIPAddress(string $humanReadableIPAddress): string
    {
        $result = inet_pton($humanReadableIPAddress);

        if ($result == false) {
            return $this->packedIPAddress('::0');
        }
        // IPv6 native
        elseif (strlen($result) == 16) {
            return $result;
        }
        // IPv4, expanded to IPv6 compatible length
        else {
            return "\0\0\0\0\0\0\0\0\0\0\0\0" . $result;
        }
    }

    /**
     * Returns true if a packed IP address is within the specified address range
     * @param string $packedAddress
     * @param string $range
     * @return bool
     * @noinspection SpellCheckingInspection
     */
    private function inIPAddressRange(string $packedAddress, string $range): bool
    {
        if ($range === 'private') {  // RFC 6890, RFC 4193
            return ($this->inIPAddressRange($packedAddress, "10.0.0.0-10.255.255.255")
                || $this->inIPAddressRange($packedAddress, "172.16.0.0-172.31.255.255")
                || $this->inIPAddressRange($packedAddress, "192.168.0.0-192.168.255.255")
                || $this->inIPAddressRange($packedAddress, "fc00::-fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"));
        } elseif ($range === 'loopback') {  // RFC 6890
            return ($this->inIPAddressRange($packedAddress, "127.0.0.1-127.255.255.255")
                || $this->inIPAddressRange($packedAddress, "::1-::1"));
        } elseif ($range === 'link-local') {  // RFC 6890, RFC 4291
            return ($this->inIPAddressRange($packedAddress, "169.254.0.0-169.254.255.255")
                || $this->inIPAddressRange($packedAddress, "fe80::-febf:ffff:ffff:ffff:ffff:ffff:ffff:ffff"));
        } else {
            $rangeLimits = explode('-', $range);
            if (count($rangeLimits) == 2) {
                $lowerLimit = $this->packedIPAddress($rangeLimits[0]);
                $upperLimit = $this->packedIPAddress($rangeLimits[1]);
                return $lowerLimit <= $packedAddress && $packedAddress <= $upperLimit;
            }
        }

        return false;
    }
}
